mindmaps-graql/src/main/java/io/mindmaps/graql/internal/analytics/ShortestPathVertexProgram.java

```diff
@@ -0,0 +1,229 @@
+                        // the assertion can also be a role-player, so sending 1 to castings following incoming edges
+                        messenger.sendMessage(messageScopeIn, MESSAGE_FROM_ROLE_PLAYER);
+                        // send -1 to castings following outgoing edges
+                        messenger.sendMessage(messageScopeOut, MESSAGE_FROM_ASSERTION);
+                    }
+                    // send message from the starting vertex
+                    String id = vertex.value(Schema.ConceptProperty.ITEM_IDENTIFIER.name());
+                    if (persistentProperties.get(START_ID).equals(id)) {
+                        LOGGER.debug("Found starting vertex");
+                        messenger.sendMessage(messageScopeIn, id);
+                        messenger.sendMessage(messageScopeOut, id);
+                    }
+                }
+                break;
+            case 1:
+                if (vertex.label().equals(Schema.BaseType.CASTING.name())) {
+//                    String id = vertex.value(Schema.ConceptProperty.ITEM_IDENTIFIER.name());
+                    Set<String> messageSet = new HashSet<>();
+                    boolean hasBothMessages = false;
+                    String fromVertex = null;
+                    Iterator<String> iterator = messenger.receiveMessages();
+                    while (iterator.hasNext()) {
+                        String message = iterator.next();
+                        if (message.length() == 1) messageSet.add(message);
+                        else fromVertex = message;
+                        if (messageSet.size() == 2) {
+                            hasBothMessages = true;
+                        }
+                        if (hasBothMessages && fromVertex != null) {
+//                            vertex.property(FROM_VERTEX, fromVertex);
+                            messenger.sendMessage(messageScopeIn, fromVertex);
+                            messenger.sendMessage(messageScopeOut, fromVertex);
+                            break;
+                        }
+                    }
+                    // casting is active if both its assertion and role-player is in the subgraph
+                    vertex.property(IS_ACTIVE_CASTING, hasBothMessages);
+                }
+                break;
+            default:
+                if (memory.get(FOUND_DESTINATION)) {
+                    String id = vertex.value(Schema.ConceptProperty.ITEM_IDENTIFIER.name());
+                    if (memory.get(NEXT_DESTINATION).equals(id)) {
+                        LOGGER.debug("Traversing back to vertex " + id);
+                        memory.set(NEXT_DESTINATION, vertex.value(FROM_VERTEX));
+                        vertex.property(FOUND_IN_ITERATION, memory.getIteration());
+                    }
+                } else {
+                    // split the default case because shortcut edges cannot be filtered out
+                    if (memory.getIteration() % 2 == 0) {
```

Comment:
I know this is probably overkill but we could do this check only once at the worker iteration start